### PR TITLE
[REVIEW] Add PyArrow based parquet readers / writers to Python, fix category dtype handling, fix arrow ingest buffer size issues

### LINKED ISSUE #896: [BUG] columnops.as_column can create incorrectly sized columns from arrow arrays
Unsure of exactly what situations cause it, but PyArrow arrays can have buffers that extend before / after the array so we need to use the `offset` and `length` attributes of the PyArrow arrays to ensure we're getting the correct data from the data buffer.

Being fixed as part of #893.

### LINKED ISSUE #910: [BUG] gdf_column_concat returns inconsistent results when given a single input column
Still working to create a minimal reproducer, but seeing this behavior before and after the libcudf call specifically where the input is a single object.

```
Arrow
=========================
cats: ('q', 'i', 'G', 'H', 't', 'j', 'e')
indices: [0, 1, 1, 1, 2, 3, 2, 4, 5, 6]

q, i, i, i, G, H, G, t, j, e


cuDF Pre-concat
=========================
cats: ('H', 'G', 'i', 'j', 'q', 'e', 't')
indices: [4 2 2 2 1 0 1 6 3 5]

q, i, i, i, G, H, G, t, j, e


cuDF Post-concat
=========================
cats: ('H', 'G', 'i', 'j', 'q', 'e', 't')
indices: [4 2 2 0 6 3 5 0 0 0]

q, i, i, H, t, j, e, H, H, H
```

### PR DESCRIPTION
**Pull request by kkraus14**
Adds cuDF API to read/write file formats using PyArrow / Pandas.

Fixes #896 
Fixes #910

### CODE REVIEW

**Comment by kkraus14** [comment]
Fixes #896

File: python/cudf/dataframe/columnops.py [hunk 1]
```diff
@@ -241,12 +241,16 @@ def as_column(arbitrary, nan_as_null=True):
             )
         elif isinstance(arbitrary, pa.DictionaryArray):
             if arbitrary.buffers()[0]:
-                pamask = Buffer(np.array(arbitrary.buffers()[0]))
+                pamask = Buffer(
+                    np.array(arbitrary.buffers()[0]).view('int8')
```
Comment (thomcom) [inline python/cudf/dataframe/columnops.py]: why is the 0th element of `arbitrary.buffers()` an `int8`?

File: python/cudf/dataframe/dataframe.py [hunk 2]
```diff
@@ -1945,6 +1946,29 @@ def quantile(self,
...
+            If ``True``, include the dataframe's index(es) in the file output.
+            If ``False``, they will not be written to the file. If ``None``,
+            the engine's default behavior will be used.
+        partition_cols : list, optional, default None
+            Column names by which to partition the dataset
+            Columns are partitioned in the order they are given
+        """
+        import cudf.io.parquet as pq
+        pq.to_parquet(self, path, compression='snappy', index=None,
+                      partition_cols=None, **kwargs)
```
[the remark this answers is not in the document]
Comment (kkraus14) [reply python/cudf/dataframe/dataframe.py]: Yep, woops, great catch.

File: python/cudf/dataframe/columnops.py [hunk 3]
```diff
@@ -291,29 +303,36 @@ def as_column(arbitrary, nan_as_null=True):
...
             else:
                 pamask = None
-            padata = Buffer(np.array(arbitrary.buffers()[1]).view(
-                np.dtype(arbitrary.type.to_pandas_dtype())
-            ))
+            padata = Buffer(
+                np.array(arbitrary.buffers()[1]).view(
+                    arbitrary.type.to_pandas_dtype()
+                )[arbitrary.offset:arbitrary.offset + len(arbitrary)]
+            )
```
[the remark this answers is not in the document]
Comment (kkraus14) [reply python/cudf/dataframe/columnops.py]: Agreed, will make it a method and use it. This also causes "unnecessary" memory copies due to https://github.com/numba/numba/issues/2521 and I'd love to change it to use `np.frombuffer` but can't quite yet 😞.

**Comment by kkraus14** [comment]
rerun tests

File: python/cudf/tests/test_parquet.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+@pytest.fixture
+def pdf():
+    return pd.DataFrame({'x': range(10),
+                         'y': range(10)})
+
+
+@pytest.fixture
+def gdf(pdf):
+    return cudf.DataFrame.from_pandas(pdf)
```
[the remark this answers is not in the document]
Comment (kkraus14) [reply python/cudf/tests/test_parquet.py]: Out of fear of breaking a whole bunch of other tests that rely on fixtures like these I'm going to keep them here for now and we can refactor into a standard used across all tests in the future.

File: CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,7 @@
 - PR #807 CSV Reader: Add byte_range parameter to specify the range in the input file to be read
 - PR #858 Add series feature hashing support
 - PR #871 CSV Reader: Add support for NA values, including user specified strings
+- PR #893 Adds Parquet, ORC, Feather, JSON, and HDF5 IO support via Pandas / PyArrow
```
Comment (thomcom) [inline CHANGELOG.md]: You just said in Slack parquet only?

File: python/cudf/dataframe/dataframe.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cudf/dataframe/dataframe.py]: You marked this resolved but it still looks hard coded to me. I just refreshed.

File: python/cudf/tests/test_parquet.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thomcom) [reply python/cudf/tests/test_parquet.py]: 👍

File: python/cudf/tests/test_parquet.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+
+@pytest.fixture(params=['snappy', 'gzip', 'botli', None])
+def parquet_file(request, tmp_path_factory, pdf):
+    fname = tmp_path_factory.mktemp("parquet") / "test.parquet"
+    pdf.to_parquet(fname, engine='pyarrow', compression=request.param)
```
Comment (thomcom) [inline python/cudf/tests/test_parquet.py]: This is `snappy` every time because of above.

File: CHANGELOG.md (hunk 5 above)
Comment (kkraus14) [inline CHANGELOG.md]: Need to update this before merge.

Comment (kkraus14) [reply CHANGELOG.md]: Will resolve.

File: python/cudf/tests/test_parquet.py (hunk 6 above)
Comment (kkraus14) [reply python/cudf/tests/test_parquet.py]: I'm not following here, the pytests show the compression param is being picked up properly based on the test names.

**Pull request merged**